src/util/PriorityQueue.h

```diff
@@ -0,0 +1,416 @@
+/**
+ * @copyright 2020, University of Freiburg, Chair of Algorithms and Data
+ * @author Johannes Kalmbach (johannes.kalmbach@gmail.com)
+ * @file Contains two implementations of Priority Queues that support updateKey
+ * operations Both of them can be used as the first template parameter to the
+ * ad_utility::FlexibleCache class template
+ *
+ * HeapBasedPQ is based on a heap(std::priority_queue). UpdateKey will cause
+ * duplicates in the internal data structure which might make later pop
+ * operations more expensive
+ *
+ * TreeBasedPQ is based on a balanced tree (std::multiset). It has stronger
+ * asymptotic guarantees, but the constant factor of the log(n) operations might
+ * be bigger because of the more complex tree structure
```

Comment:
I'm really curious what the real world performance of this will be. My gut feeling is that with todays CPUs the bad caching behaviour of (non-B) Trees will make this fare significantly worse than the HeapBasedPQ

Reply:
That might be true, but it might also be, that we only access relatively expensive cache Entries (rather long std::vectors) so the actual access time to the cache does not really matter.